File: src/main/java/io/jjong/algorithm/Parity.java
package io.jjong.algorithm;

/**
 * create on 2022/12/09. create by IntelliJ IDEA.
 *
 * <p> 클래스 설명 </p>
 *
 * @author Jongsang Han(henry)
 * @version 1.0
 * @see
 * @since 1.0
 */
public class Parity {

  public static void main(String[] args) {
    long x = 220;
    short parity = parity(x);
    System.out.println("parity : " + parity);

//    int x = 27;   // -64;
//    System.out.println(Integer.toBinaryString(x));
//    int n = 2;
//    System.out.println("x >> n = " + (x >> n)); // 나눗셈, (x / 2^n)
//    System.out.println("x << n = " + (x << n)); // 곱셈,  (x * 2^n)
//    System.out.println("x >>> n = " + (x >>> n));
//
//    System.out.println("x >> 34 = " + (x >> 2));    // x / 2^(n % 32) 0001 1111
//    System.out.println(Integer.toBinaryString(x >>> 2));
//    System.out.println("x << 34 = " + (x << 3));
//    System.out.println("x >>> 34 = " + (x >>> 3));
  }

  /**
   * 에러를 검출하기 위해 추가하는 비트
   * @param x
   * @return
   */
  public static short parity(long x) {
    short result = 0;
    while (x != 0) {
      result ^= (x & 1); // x가 1인지 체크 한뒤 xor 연산 수행.
      System.out.println("x & 1 : "+ (x & 1)+", result : " + result);
      x >>>= 1; // 나누기 연산 수행 -> x / 2^1
      System.out.println("x : " + x);
    }
    return result;
  }

}

File: src/main/kotlin/io/jjong/algorithm/Parity.kt
package io.jjong.algorithm

/**
 * 비트 패리티(1인 비트 개수의 홀짝)를 반환한다. 1이 홀수 개면 1, 짝수 개면 0.
 * 최하위 비트를 xor로 누적하며 우측 시프트로 비트를 하나씩 소거한다.
 */
object Parity {
    fun parity(x: Long): Short {
        var n = x
        var result = 0 // Short xor 연산은 Int를 반환하므로 Int로 누적 후 변환한다.
        while (n != 0L) {
            result = result xor (n and 1L).toInt() // 최하위 비트를 검사해 xor 누적
            println("x & 1 : ${n and 1L}, result : $result")
            n = n ushr 1 // x / 2^1
            println("x : $n")
        }
        return result.toShort()
    }
}

fun main() {
    val x = 220L
    val parity = Parity.parity(x)
    println("parity : $parity")
}
